### PR TITLE
Fixing swagger correctness items for RecoveryServicesBackup[s360]

### PR DESCRIPTION
**Pull request by pratimaupadhyay02**
<i>MSFT employees can try out our new experience at <b>[OpenAPI Hub](https://aka.ms/openapiportal) </b> - one location for using our validation tools and finding your workflow. 
</i>

### Changelog
Please ensure to add changelog with this PR by answering the following questions.
  1. What's the purpose of the update?    
      - [ ] new service onboarding 
      - [x] new API version 
      - [ ] update existing version for new feature 
      - [x] update existing version to fix swagger quality issue in s360
      - [ ] Other, please clarify 
  2. When you are targeting to deploy new service/feature to public regions? Please provide date, or month to public if date is not available yet.
  3. When you expect to publish swagger? Please provide date, or month to public if date is not available yet.
  4. If it's an update to existing version,  please select SDKs of specific language and CLIs that require refresh after swagger is published.
      - [ ] SDK of .NET (need service team to ensure code readiness)
      - [ ] SDK of Python
      - [ ] SDK of Java
      - [ ] SDK of Js
      - [ ] SDK of Go
      - [ ] PowerShell
      - [ ] CLI
      - [ ] Terraform
      - [ ] No, no need to refresh for updates in this PR

### Contribution checklist:
- [ ] I commit to follow the [Breaking Change Policy](http://aka.ms/bcforapi) of "no breaking changes"
- [ ] I have reviewed the [documentation](https://aka.ms/ameonboard) for the workflow.
- [ ] [Validation tools](https://aka.ms/swaggertools) were run on swagger spec(s) and errors have all been fixed in this PR. [How to fix?](https://aka.ms/ci-fix)

If any further question about AME onboarding or validation tools, please view the [FAQ](https://aka.ms/faqinprreview).

### ARM API Review Checklist
- [x] Ensure to check this box if one of the following scenarios meet updates in the PR, so that label “WaitForARMFeedback” will be added automatically to involve ARM API Review. Failure to comply may result in delays for manifest application. Note this does not apply to data plane APIs, all “removals” and “adding a new property” no more require ARM API review.
  - Adding new API(s)
  - Adding a new API version
  -  [ ] Ensure to copy the existing version into new directory structure for first commit (including refactoring) and then push new changes including version updates in separate commits. This is required to review the changes efficiently.
  - Adding a new service

- [ ] Please ensure you've reviewed following [guidelines](https://aka.ms/rpguidelines) including [ARM resource provider contract](https://github.com/Azure/azure-resource-manager-rpc) and [REST guidelines](https://github.com/microsoft/api-guidelines/blob/vNext/azure/Guidelines.md). Estimated time (4 hours). This is required before you can request review from ARM API Review board.

- [ ] If you are blocked on ARM review and want to get the PR merged with urgency, please get the ARM oncall for reviews (*RP Manifest Approvers* team under <ins>Azure Resource Manager service</ins>) from IcM and reach out to them. 

### Breaking Change Review Checklist 
If there are following updates in the PR, ensure to request an approval from Breaking Change Review Board as defined in the [Breaking Change Policy](http://aka.ms/bcforapi). 

- [ ] Removing API(s) in stable version
- [ ] Removing properties in stable version
- [ ] Removing API version(s) in stable version
- [ ] Updating API in stable or public preview version with Breaking Change Validation errors
- [ ] Updating API(s) in public preview over 1 year (refer to [Retirement of Previews](https://dev.azure.com/msazure/AzureWiki/_wiki/wikis/AzureWiki.wiki/37683/Retirement-of-Previews))

**Action**: to initiate an evaluation of the breaking change, create a new intake using the [template for breaking changes](https://aka.ms/Breakingchangetemplate). Addition details on the process and office hours are on the [Breaking change Wiki](https://dev.azure.com/msazure/AzureWiki/_wiki/wikis/AzureWiki.wiki/37684/Breaking-Changes).

Please follow the link to find more details on [PR review process](https://aka.ms/SwaggerPRReview).

### CODE REVIEW

**Comment by mentat9** [comment]
>           "204": {

PATCH response should have 200 status and the body should be the updated resource.

RPC reference: [https://github.com/Azure/azure-resource-manager-rpc/blob/master/v1.0/resource-api-reference.md#patch-resource](https://github.com/Azure/azure-resource-manager-rpc/blob/master/v1.0/resource-api-reference.md#patch-resource)

MS REST guidelines: [https://github.com/Microsoft/api-guidelines/blob/master/Guidelines.md#742-patch](https://github.com/Microsoft/api-guidelines/blob/master/Guidelines.md#742-patch)

---
Refers to: specification/recoveryservicesbackup/resource-manager/Microsoft.RecoveryServices/stable/2021-07-01/bms.json:147 in 47003ac. [](commit_id = 47003ac9f1974d811d7aba4438f32b862fbca63d, deletion_comment = False)

**Comment by mentat9** [comment]
>         "properties": {

Encoding objects in strings is not generally preferred. Can this be schematized?

---
Refers to: specification/recoveryservicesbackup/resource-manager/Microsoft.RecoveryServices/stable/2021-07-01/bms.json:9623 in 47003ac. [](commit_id = 47003ac9f1974d811d7aba4438f32b862fbca63d, deletion_comment = False)

**Comment by mentat9** [comment]
>         "responses": {

Should have a default branch for error status codes.

---
Refers to: specification/recoveryservicesbackup/resource-manager/Microsoft.RecoveryServices/stable/2021-07-01/bms.json:194 in 47003ac. [](commit_id = 47003ac9f1974d811d7aba4438f32b862fbca63d, deletion_comment = False)

**Comment by mentat9** [comment]
>         "responses": {

Need a default case.

---
Refers to: specification/recoveryservicesbackup/resource-manager/Microsoft.RecoveryServices/stable/2021-07-01/bms.json:239 in 47003ac. [](commit_id = 47003ac9f1974d811d7aba4438f32b862fbca63d, deletion_comment = False)

**Comment by mentat9** [comment]
>         "errorCode": {

Error code and message should be returned in their own schema and a separate status code.

---
Refers to: specification/recoveryservicesbackup/resource-manager/Microsoft.RecoveryServices/stable/2021-07-01/bms.json:9646 in 47003ac. [](commit_id = 47003ac9f1974d811d7aba4438f32b862fbca63d, deletion_comment = False)

**Comment by mentat9** [comment]
>         "errorCode": {

Error code and message should be returned in their own schema and a separate status code. You can $ref error definitions from common-types/resource-management/v2/types.json to follow standard schema.

---
Refers to: specification/recoveryservicesbackup/resource-manager/Microsoft.RecoveryServices/stable/2021-07-01/bms.json:9516 in 47003ac. [](commit_id = 47003ac9f1974d811d7aba4438f32b862fbca63d, deletion_comment = False)

File: specification/recoveryservicesbackup/resource-manager/Microsoft.RecoveryServices/stable/2018-12-20/bms.json
Comment (mentat9) [inline specification/recoveryservicesbackup/resource-manager/Microsoft.RecoveryServices/stable/2018-12-20/bms.json:3944]: >isArchiveEnabled [](http://example.com/codeflow?start=9&length=16)

String enums are preferred to booleans because they are easier for customers to use, allow additional values to be added, and can be designed to avoid invalid combinations with other switches. For example, this property could be archiveMethod: { None, Full } or similar.

**Comment by mentat9** [comment]
>     "/Subscriptions/{subscriptionId}/resourceGroups/{resourceGroupName}/providers/Microsoft.RecoveryServices/vaults/{vaultName}/backupstorageconfig/vaultstorageconfig": {

Should also implement collection GET on this URL: /Subscriptions/{subscriptionId}/resourceGroups/{resourceGroupName}/providers/Microsoft.RecoveryServices/vaults/{vaultName}/backupstorageconfig.

Even though this is a singleton, the collection GET should return an array containing the single instance. This is required for discovery. Some features won't work without it (ARG, Azure Policy, probably others).

---
Refers to: specification/recoveryservicesbackup/resource-manager/Microsoft.RecoveryServices/stable/2021-07-01/bms.json:21 in 47003ac. [](commit_id = 47003ac9f1974d811d7aba4438f32b862fbca63d, deletion_comment = False)

**Comment by mentat9** [comment]
>         "responses": {

Should return 200 if the resource was deleted. Also need default case for errors.

---
Refers to: specification/recoveryservicesbackup/resource-manager/Microsoft.RecoveryServices/stable/2021-07-01/bms.json:448 in 47003ac. [](commit_id = 47003ac9f1974d811d7aba4438f32b862fbca63d, deletion_comment = False)

**Comment by mentat9** [comment]
>     "/subscriptions/{subscriptionId}/resourceGroups/{resourceGroupName}/providers/Microsoft.RecoveryServices/vaults/{vaultName}/backupconfig/vaultconfig": {

Should also add collection GET on the type for discovery of the singleton instance.

---
Refers to: specification/recoveryservicesbackup/resource-manager/Microsoft.RecoveryServices/stable/2021-07-01/bms.json:608 in 47003ac. [](commit_id = 47003ac9f1974d811d7aba4438f32b862fbca63d, deletion_comment = False)

**Comment by mentat9** [comment]
>     "/subscriptions/{subscriptionId}/resourceGroups/{resourceGroupName}/providers/Microsoft.RecoveryServices/vaults/{vaultName}/backupEncryptionConfigs/backupResourceEncryptionConfig": {

Also should implement collection GET for discovery.

---
Refers to: specification/recoveryservicesbackup/resource-manager/Microsoft.RecoveryServices/stable/2021-07-01/bms.json:757 in 47003ac. [](commit_id = 47003ac9f1974d811d7aba4438f32b862fbca63d, deletion_comment = False)

**Comment by mentat9** [comment]
>             "description": "Created",

If this isn't the async response, you need to add 202 for async response.

---
Refers to: specification/recoveryservicesbackup/resource-manager/Microsoft.RecoveryServices/stable/2021-07-01/bms.json:949 in 47003ac. [](commit_id = 47003ac9f1974d811d7aba4438f32b862fbca63d, deletion_comment = False)

**Comment by mentat9** [comment]
>     "/subscriptions/{subscriptionId}/resourceGroups/{resourceGroupName}/providers/Microsoft.RecoveryServices/vaults/{vaultName}/privateEndpointConnections/{privateEndpointConnectionName}/operationsStatus/{operationId}": {

Should also implement collection GET for discovery.

---
Refers to: specification/recoveryservicesbackup/resource-manager/Microsoft.RecoveryServices/stable/2021-07-01/bms.json:1023 in 47003ac. [](commit_id = 47003ac9f1974d811d7aba4438f32b862fbca63d, deletion_comment = False)

**Comment by mentat9** [comment]
>     "/subscriptions/{subscriptionId}/resourceGroups/{resourceGroupName}/providers/Microsoft.RecoveryServices/vaults/{vaultName}/privateEndpointConnections/{privateEndpointConnectionName}/operationsStatus/{operationId}": {

Should also implement collection GET for discovery.

---
Refers to: specification/recoveryservicesbackup/resource-manager/Microsoft.RecoveryServices/stable/2021-07-01/bms.json:1023 in 47003ac. [](commit_id = 47003ac9f1974d811d7aba4438f32b862fbca63d, deletion_comment = False)

**Comment by mentat9** [comment]
>     "/subscriptions/{subscriptionId}/resourceGroups/{resourceGroupName}/providers/Microsoft.RecoveryServices/vaults/{vaultName}/backupstorageconfig/vaultstorageconfig/operationStatus/{operationId}": {

Should implement collection GET for discovery.

---
Refers to: specification/recoveryservicesbackup/resource-manager/Microsoft.RecoveryServices/stable/2021-07-01/bms.json:1082 in 47003ac. [](commit_id = 47003ac9f1974d811d7aba4438f32b862fbca63d, deletion_comment = False)

**Comment by mentat9** [comment]
>           "202": {

Need to add x-ms-long-running-operation: true.

---
Refers to: specification/recoveryservicesbackup/resource-manager/Microsoft.RecoveryServices/stable/2021-07-01/bms.json:1500 in 47003ac. [](commit_id = 47003ac9f1974d811d7aba4438f32b862fbca63d, deletion_comment = False)

**Comment by mentat9** [comment]
>         "responses": {

Should also return 201 for create.

---
Refers to: specification/recoveryservicesbackup/resource-manager/Microsoft.RecoveryServices/stable/2021-07-01/bms.json:97 in 47003ac. [](commit_id = 47003ac9f1974d811d7aba4438f32b862fbca63d, deletion_comment = False)

**Comment by mentat9** [comment]
>         "responses": {

Should also return 201 for create.

---
Refers to: specification/recoveryservicesbackup/resource-manager/Microsoft.RecoveryServices/stable/2021-07-01/bms.json:397 in 47003ac. [](commit_id = 47003ac9f1974d811d7aba4438f32b862fbca63d, deletion_comment = False)

**Comment by mentat9** [comment]
>         "responses": {

Should also return 201 for create.

---
Refers to: specification/recoveryservicesbackup/resource-manager/Microsoft.RecoveryServices/stable/2021-07-01/bms.json:736 in 47003ac. [](commit_id = 47003ac9f1974d811d7aba4438f32b862fbca63d, deletion_comment = False)

**Comment by mentat9** [comment]
>         "responses": {

Should also return 201 for create.

---
Refers to: specification/recoveryservicesbackup/resource-manager/Microsoft.RecoveryServices/stable/2021-07-01/bms.json:833 in 47003ac. [](commit_id = 47003ac9f1974d811d7aba4438f32b862fbca63d, deletion_comment = False)

**Comment by mentat9** [comment]
>           "202": {

operationResults should not be async or return 202.

---
Refers to: specification/recoveryservicesbackup/resource-manager/Microsoft.RecoveryServices/stable/2021-07-01/bms.json:1225 in 47003ac. [](commit_id = 47003ac9f1974d811d7aba4438f32b862fbca63d, deletion_comment = False)

**Comment by mentat9** [comment]
>     "/subscriptions/{subscriptionId}/resourceGroups/{resourceGroupName}/providers/Microsoft.RecoveryServices/vaults/{vaultName}/backupstorageconfig/vaultstorageconfig/operationResults/{operationId}": {

ARM recommends modeling operationResults as a root-level resource.

Wiki link: [https://armwiki.azurewebsites.net/rp_onboarding/process/api_review_best_practices.html?q=operationResults#common-issues-found-in-review](https://armwiki.azurewebsites.net/rp_onboarding/process/api_review_best_practices.html?q=operationResults#common-issues-found-in-review)

---
Refers to: specification/recoveryservicesbackup/resource-manager/Microsoft.RecoveryServices/stable/2021-07-01/bms.json:1188 in 47003ac. [](commit_id = 47003ac9f1974d811d7aba4438f32b862fbca63d, deletion_comment = False)

**Comment by mentat9** [comment]
>         "responses": {

Need to add 201 for create case

---
Refers to: specification/recoveryservicesbackup/resource-manager/Microsoft.RecoveryServices/stable/2021-07-01/bms.json:1426 in 47003ac. [](commit_id = 47003ac9f1974d811d7aba4438f32b862fbca63d, deletion_comment = False)

**Comment by mentat9** [comment]
>           "202": {

Need to add x-ms-long-running-operation: true.

---
Refers to: specification/recoveryservicesbackup/resource-manager/Microsoft.RecoveryServices/stable/2021-07-01/bms.json:1579 in 47003ac. [](commit_id = 47003ac9f1974d811d7aba4438f32b862fbca63d, deletion_comment = False)

**Comment by mentat9** [comment]
>         "crossRegionRestoreFlag": {

String enums are encouraged over boolean. This could be something like restoreType: { Default, CrossRegion }.

---
Refers to: specification/recoveryservicesbackup/resource-manager/Microsoft.RecoveryServices/stable/2021-07-01/bms.json:4965 in 47003ac. [](commit_id = 47003ac9f1974d811d7aba4438f32b862fbca63d, deletion_comment = False)

**Comment by mentat9** [comment]
>         "isArchiveEnabled": {

Consider using string enum instead of boolean, e.g. archiveType: { None, Full }, or similar.

---
Refers to: specification/recoveryservicesbackup/resource-manager/Microsoft.RecoveryServices/stable/2021-07-01/bms.json:7702 in 47003ac. [](commit_id = 47003ac9f1974d811d7aba4438f32b862fbca63d, deletion_comment = False)

**Comment by mentat9** [comment]
>         "isSoftDeleteFeatureStateEditable": {

Could be enum, e.g. softDeleteSupport: { ReadOnly, ReadWrite }, or similar. Could also possibly implement this by adding a value to the softDeleteFeatureState enum, e.g. EnabledPermanently, or EnabledEditable, or similar (whichever new case is being added).

---
Refers to: specification/recoveryservicesbackup/resource-manager/Microsoft.RecoveryServices/stable/2021-07-01/bms.json:8655 in 47003ac. [](commit_id = 47003ac9f1974d811d7aba4438f32b862fbca63d, deletion_comment = False)

**Comment by mentat9** [comment]
The checklist for this PR was not filled out correctly: a new API version is being added.

Also, two requirements were not fulfilled:
1. When adding an API version, it is very important to follow this requirement: **Ensure to copy the existing version into new directory structure for first commit (including refactoring) and then push new changes including version updates in separate commits. This is required to review the changes efficiently.**
2. Please ensure you've reviewed following guidelines including ARM resource provider contract and REST guidelines. Estimated time (4 hours). This is required before you can request review from ARM API Review board.

I didn't notice these issues until I had already invested a bunch of time, so I completed the initial review pass. Please study the guidelines before your next PR, and carefully fill in the checklist when submitting.

**Comment by pratimaupadhyay02** [comment]
Reverted the changes since these were causing Model Validation Failures. As suggested by @PhoenixHe-msft , we will split our RP into multiple packages containing CRR and non CRR API versions to avoid these issues in the future.

**Comment by pratimaupadhyay02** [comment]
Currently the PR is only combining the change in previous 2021-06-01 API version and 2021-04-01 version in a new version 2021-07-01 version. No incremental changes are done here. Do we require ARM review in this case?

**Comment by pratimaupadhyay02** [comment]
Hi @mentat9, since our RP contains multiple API versions under a single package, fixing few correctness Items was causing ModelValidation failures due to which I have reverted the previous set of changes. We can take these changes forward only after splitting our RP into multiple packages. Currently the PR only contains combining already existing changes from 2021-04-01 and 2021-06-01 version into a new version 2021-07-01 for SDK generation purposes. Can you please review the PR.

**Comment by mentat9** [comment]
@pratimaupadhyay02: you will be working with this week's oncall (@majastrz) to continue your review.

**Comment by pratimaupadhyay02** [comment]
@majastrz Please review the PR

**Comment by majastrz** [comment]
>     "/Subscriptions/{subscriptionId}/resourceGroups/{resourceGroupName}/providers/Microsoft.RecoveryServices/vaults/{vaultName}/backupstorageconfig/vaultstorageconfig": {

This hasn't been addressed.

---
In reply to: [881268189](https://github.com/Azure/azure-rest-api-specs/pull/15139#issuecomment-881268189) [](http://example.com/codeflow?ancestors=881268189,881268189)

---
Refers to: specification/recoveryservicesbackup/resource-manager/Microsoft.RecoveryServices/stable/2021-07-01/bms.json:21 in 47003ac. [](commit_id = 47003ac9f1974d811d7aba4438f32b862fbca63d, deletion_comment = False)

**Comment by majastrz** [comment]
>           "204": {

This hasn't been addressed.

---
In reply to: [881247377](https://github.com/Azure/azure-rest-api-specs/pull/15139#issuecomment-881247377) [](http://example.com/codeflow?ancestors=881247377,881247377)

---
Refers to: specification/recoveryservicesbackup/resource-manager/Microsoft.RecoveryServices/stable/2021-07-01/bms.json:147 in 47003ac. [](commit_id = 47003ac9f1974d811d7aba4438f32b862fbca63d, deletion_comment = False)

**Comment by majastrz** [comment]
@mentat9's comments have not been addressed. I replied to each one to point them out. Please take a look.

**Comment by pratimaupadhyay02** [comment]
> @mentat9's comments have not been addressed. I replied to each one to point them out. Please take a look.

Hi @majastrz, this is an already existing API introduced in 2018. The same format is used in 2018-12-20 which is under the same package. Making changes at this stage will break the existing code. Please find the already existing API in 2018-12-20 https://github.com/Azure/azure-rest-api-specs/blob/552b277f9f8d2057a94dfc9c3e09c0d877c0ea30/specification/recoveryservicesbackup/resource-manager/Microsoft.RecoveryServices/stable/2018-12-20/bms.json#L491. It is being added here since 2018-12-20 version is reserved for cross region restore APIs so the call is routed to passive stamp based on the version. For calls to be routed to active stamp, the API should be present in the latest version so that is the reason of adding it here.

**Review by j5lim — APPROVED** [review]
Combining two PRs approved for s360. Approving from ARM side. Please address other reviewers' comments in a next version.

**Comment by pratimaupadhyay02** [comment]
[Do not merge] These changes are merged in another PR. I am keeping this PR open to test a few scenarios in readme.md

**Pull request closed without merging**
